### PR TITLE
Add generative models and datasets

### PR DESCRIPTION
**Pull request by josephbowles**
Major update to the repo to include new generative models and datasets. 

**New Models**
- RestrictedBoltzmannMachine: A wrapped version of sklearn's BernoulliRBM model 
- DeepEBM: an energy based model in JAX that uses a MLP network as its energy function and is trained with k-contrastive divergence. 

Both models can be scored using the maximum mean discrepancy distance. 

**New dataset generating functions**

These are designed for generative learning
- generate_ising: sample data from an ising distribution via Metropolis Hastings MCMC
- generate_spin_blobs: a binary version of 'gaussian blobs' where configurations are sampled close in Hamming weight to a set of peak configuration

**Backwards compatibility**
All code from the previous version of the repo should work identically. Training and hyperparameter optimization scripts have been updated accordingly to handle generative models. As a general rule, we specify the model is generative by using y=None in place of labels. 

**future changes**
The repo does not contain any quantum generative models at the moment; it would be good to add a simple example. There is also no changes to the folder structure, so e.g. classifier and generative models are stored in the same directory. It may make sense to change this structure to account for the classifier/generative dichotomy going forward.

### CODE REVIEW

File: scripts/score_with_best_hyperparameters.py [hunk 1]
```diff
@@ -12,7 +12,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Score a model using the best hyperparameters, using a command-line script."""
+"""
+Score a model using the best hyperparameters, using a command-line script.
+Note this is only compatible with classifier models.
```
Comment (mariaschuld) [inline scripts/score_with_best_hyperparameters.py:17]:
```suggestion
Note this is only compatible with supervised models for classification.
```

(Minor)

File: src/qml_benchmarks/data/ising.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                   J: jnp.array,
+                   b: jnp.array,
+                   T: float,
+                   sparse=False,
+                   num_chains=1,
+                   thinning=1,
+                   num_warmup=1000,
+                   key=42):
+    r"""
+    Generating function for ising datasets.
```
Comment (mariaschuld) [inline src/qml_benchmarks/data/ising.py:199]:
```suggestion
    Generating function for Ising datasets.
```
Minor, but looks more professional to use the right capitalisation?

File: src/qml_benchmarks/data/ising.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    The dataset is generated by sampling an ising distrbution of a specified interaction
+    matrix. The distribution is sampled via markov chain Monte Carlo via the Metrolopis Hastings
```
Comment (mariaschuld) [inline src/qml_benchmarks/data/ising.py:202]:
```suggestion
    matrix. The distribution is sampled via Markov Chain Monte Carlo via the Metrolopis Hastings
```

**Review by mariaschuld — APPROVED** [review]
Only minor comments, but overall good to go! Thanks for making the changes, and nice to see that it wasn't overly painful to extend the scope of the repo...
